docs/proposals/20200506-conditions.md

```diff
@@ -0,0 +1,363 @@
+type ConditionSeverity string
+
+const (
+  // ConditionSeverityError specifies that a failure of a condition type
+  // should be viewed as an error.
+  ConditionSeverityError ConditionSeverity = "Error"
+
+  // ConditionSeverityWarning specifies that a failure of a condition type
+  // should be viewed as a warning, but that things could still work.
+  ConditionSeverityWarning ConditionSeverity = "Warning"
+
+  // ConditionSeverityInfo specifies that a failure of a condition type
+  // should be viewed as purely informational, and that things could still work.
+  ConditionSeverityInfo ConditionSeverity = "Info"
+
+  // ConditionSeverityNone should apply only if the condition is in state "True".
+  // As ConditionSeverityNone is the default for conditions we use the empty string (coupled with omitempty)
+  ConditionSeverityNone ConditionSeverity = ""
+)
+
+// Conditions define a readiness condition for a Cluster API resource.
+type Condition struct {
+   // Type of condition.
+   // +required
+   Type ConditionType `json:"type" description:"type of status condition"`
+
+   // Status of the condition, one of True, False, Unknown.
+   // +required
+   Status corev1.ConditionStatus `json:"status"`
+
+   // Severity with which to treat failures of this type of condition.
+   // When this is not specified, it defaults to Error.
+   // +optional
+   Severity ConditionSeverity `json:"severity,omitempty"`
+
+   // LastTransitionTime is the last time the condition transitioned from one status to another.
+   // +optional
+   LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+   // The reason for the condition's last transition.
+   // Reasons should be one-word CamelCase.
+   // +optional
+   Reason string `json:"reason,omitempty" description:"one-word CamelCase reason for the condition's last transition"`
+
+   // A human readable message indicating details about the transition.
+   // +optional
+   Message string `json:"message,omitempty" description:"human-readable message indicating details about last transition"`
+}
+
+// Conditions is the schema for the conditions portion of a Cluster API resource.
+type Conditions []Condition
+```
+
+Every Cluster API resource except data-only objects which aren't reconciled (i.e. `KubeadmConfigTemplate`)
+SHOULD have conditions field in the status struct. e.g.
+
+```golang
+// ClusterStatus represent the status for the Cluster object
+type ClusterStatus struct {
+    ...
+    // Conditions define a list of readiness conditions for the Cluster object
+    Conditions Conditions `json:"conditions,omitempty"
+} 
+```
+
+Each Cluster API resource SHOULD define its own set of condition types, e.g. 
+
+```golang
+// ConditionTypes for the cluster object 
+const (
+	ClusterInfrastructureReady ConditionType = "ClusterInfrastructureReady"
+	
+	...
+)
+```
+
+Condition types MUST have a consistent polarity (i.e. "True = good");
+
+Condition types MUST have one of the following suffix:
```

Comment:
How would we express things like `Timeouts`, for example from MachineHealthCheck

Reply:
IMO a missed timeout is a reason for a condition failing, not a separate condition